specreduce/background.py

```diff
@@ -0,0 +1,205 @@
+            image with 2-D spectral image data
+        trace_object: Trace
+            estimated trace of the spectrum to center the background traces
+        separation: float
+            separation from ``trace_object`` for the background regions
+        width : float
+            width of each background aperture in pixels
+        statistic: string
+            statistic to use when computing the background.  'average' will
+            account for partial pixel weights, 'median' will include all partial
+            pixels.
+        disp_axis : int
+            dispersion axis
+        crossdisp_axis : int
+            cross-dispersion axis
+        """
+        kwargs['traces'] = [trace_object-separation, trace_object+separation]
+        return cls(image=image, **kwargs)
+
+    @classmethod
+    def one_sided(cls, image, trace_object, separation, **kwargs):
+        """
+        Determine the background from an image for subtraction above
+        or below an input trace.
+
+        Parameters
+        ----------
+        image : nddata-compatible image
+            image with 2-D spectral image data
+        trace_object: Trace
+            estimated trace of the spectrum to center the background traces
+        separation: float
+            separation from ``trace_object`` for the background, positive will be
+            above the trace, negative below.
+        width : float
+            width of each background aperture in pixels
+        statistic: string
+            statistic to use when computing the background.  'average' will
+            account for partial pixel weights, 'median' will include all partial
+            pixels.
+        disp_axis : int
+            dispersion axis
+        crossdisp_axis : int
+            cross-dispersion axis
+        """
+        kwargs['traces'] = [trace_object+separation]
+        return cls(image=image, **kwargs)
+
+    def bkg_image(self, image=None):
+        """
+        Expose the background tiled to the dimension of ``image``.
+
+        Parameters
+        ----------
+        image : nddata-compatible image or None
+            image with 2-D spectral image data.  If None, will use ``image`` passed
+            to extract the background.
+
+        Returns
+        -------
+        array with same shape as ``image``.
+        """
+        if image is None:
+            image = self.image
+
+        return np.tile(self.bkg_array, (image.shape[0], 1))
+
+    def sub_image(self, image=None):
+        """
+        Subtract the computed background from ``image``.
+
+        Parameters
+        ----------
+        image : nddata-compatible image or None
+            image with 2-D spectral image data.  If None, will use ``image`` passed
+            to extract the background.
+
+        Returns
+        -------
+        array with same shape as ``image``
+        """
+        if image is None:
+            image = self.image
+
+        return image - self.bkg_image(image)
```

Comment:
this works for normal `ndarray` instances, but does not work for `NDData` and subclasses. they implement their own arithmetic methods so they can support different ways of handling masks and propagating uncertainties. see https://docs.astropy.org/en/stable/nddata/mixins/ndarithmetic.html for details.